src/crimson/os/seastore/object_data_handler.cc

```diff
@@ -284,7 +283,6 @@ ObjectDataHandler::clear_ret ObjectDataHandler::trim_data_reservation(
   context_t ctx, object_data_t &object_data, extent_len_t size)
 {
   ceph_assert(!object_data.is_null());
-  assert_aligned(size);
```

Comment:
I think this has to be aligned.

Reply:
the size is op.extent.offset getting from m->ops,  need the exact size to do bl append in  the following. so it can't be aligned here.

Reply:
This gets called from ObjectDataHandler::truncate() and ObjectDataHandler::clear().  ObjectDataHandler::clear() passes in 0 and therefore is aligned.  ObjectDataHandler::truncate() should be rounding offset up before passing to trim_data_reservation().  Let's leave the assert here and fix ObjectDataHandler::truncate().

Reply:
Hi Sam, trim_data_reservation need read data out to bufferlist, it should read the exact extent.offset, if we roundup before it, it will read out a block size aligned data, (more data than it is expected).

Reply:
see the code in trim_data_reservation: 
	    bl.append(
	      bufferptr(
		extent->get_bptr(),
		0,
		size - pin_offset       //if we round size up to block aligned, we will read more data then expected. 
	      ));

Reply:
We aren't doing a read here, we're just releasing the data extents that have been truncated.  If offset is non-aligned, we need to keep the whole block containing it.  trim_data_reservation only performs a read in order to rewrite the portion of an extent prior to the truncated size.

There may be a bug in trim_data_reservation if pins.front().get_laddr() == object_data.get_reserved_data_base() + size.  Might need:

```
--- a/src/crimson/os/seastore/object_data_handler.cc
+++ b/src/crimson/os/seastore/object_data_handler.cc
@@ -316,12 +316,18 @@ ObjectDataHandler::clear_ret ObjectDataHandler::trim_data_reservation(
          pin.get_laddr() <= object_data.get_reserved_data_base() + size);
        auto pin_offset = pin.get_laddr() -
          object_data.get_reserved_data_base();
-       if (pin.get_paddr().is_zero()) {
+       if ((pin.get_laddr() == (object_data.get_reserved_data_base() + size)) ||
+           (pin.get_paddr().is_zero())) {
+         /* First pin is exactly at the boundary or is a zero pin.  Either way,
+          * remove all pins and add a single zero pin to the end.
+          */
          to_write.emplace_back(
            pin.get_laddr(),
            object_data.get_reserved_data_len() - pin_offset);
          return clear_iertr::now();
        } else {
+         /* First pin overlaps the boundary and has data, read in extent
+          * and rewrite portion prior to size */
          return read_pin(
            ctx,
            pin.duplicate()
```

Reply:
ok, got it we keep the whole block aligned size content.  will verify it first.